pkg/util/pki/csr_test.go

```diff
@@ -553,3 +556,126 @@ func Test_buildKeyUsagesExtensionsForCertificate(t *testing.T) {
 		})
 	}
 }
+
+func TestSignCSRTemplate(t *testing.T) {
+	rootPrivKey, err := GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	rootTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(0),
+		Subject: pkix.Name{
+			CommonName: "testing-root",
+		},
+		PublicKey: rootPrivKey.Public(),
+		IsCA:      true,
+	}
+
+	rootPem, rootCert, err := SignCertificate(rootTemplate, rootTemplate, rootTemplate.PublicKey, rootPrivKey)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	interPrivKey, err := GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	interTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: "testing-intermediate",
+		},
+		PublicKey: interPrivKey.Public(),
+		IsCA:      true,
+	}
+
+	interPem, interCert, err := SignCertificate(interTemplate, rootCert, interTemplate.PublicKey, rootPrivKey)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	leafPrivKey, err := GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	leafTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(2),
+		Subject: pkix.Name{
+			CommonName: "testing-leaf",
+		},
+		PublicKey: leafPrivKey.Public(),
+	}
+
+	leafPem, _, err := SignCertificate(leafTemplate, interCert, leafTemplate.PublicKey, interPrivKey)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
```

Comment:
Your unit tests are great!!! I ran them locally, everything passes.

Suggestion: to improve readability, we could add comments to guide the eyes of the reader. I also used `assert.NoError` to fold the error blocks to one line:

```suggestion
	// We want to be testing the behavior of SignCSRTemplate in the various
	// context; for that, we construct a chain of three certificates: a root, an
	// intermediate an a leaf certificate.

	// We start with a root CA:
	rootPrivKey, err := GenerateRSAPrivateKey(2048)
	require.NoError(t, err)
	rootTemplate := &x509.Certificate{
		SerialNumber: big.NewInt(0),
		Subject: pkix.Name{
			CommonName: "testing-root",
		},
		PublicKey: rootPrivKey.Public(),
		IsCA:      true,
	}
	rootPem, rootCert, err := SignCertificate(rootTemplate, rootTemplate, rootTemplate.PublicKey, rootPrivKey)
	require.NoError(t, err)

	// Then, we want an intermediate CA:
	interPrivKey, err := GenerateRSAPrivateKey(2048)
	require.NoError(t, err)
	interTemplate := &x509.Certificate{
		SerialNumber: big.NewInt(1),
		Subject: pkix.Name{
			CommonName: "testing-intermediate",
		},
		PublicKey: interPrivKey.Public(),
		IsCA:      true,
	}
	interPem, interCert, err := SignCertificate(interTemplate, rootCert, interTemplate.PublicKey, rootPrivKey)
	require.NoError(t, err)

	// And finally, we also want a leaf certificate:
	leafPrivKey, err := GenerateRSAPrivateKey(2048)
	require.NoError(t, err)
	leafTemplate := &x509.Certificate{
		SerialNumber: big.NewInt(2),
		Subject: pkix.Name{
			CommonName: "testing-leaf",
		},
		PublicKey: leafPrivKey.Public(),
	}
	leafPem, _, err := SignCertificate(leafTemplate, interCert, leafTemplate.PublicKey, interPrivKey)
	require.NoError(t, err)
```

The import for testify:

```go
    "github.com/stretchr/testify/require"
```

Reply:
@maelvls Thanks a lot for your feedback! It already looks a LOT better with your suggestions. I will update, and shift focus over to the e2e tests. WDYT about opening a separate PR for adding these unit tests? It would then not test the behavior using intermediate. But I did NOT liked that my PR labels were changed from `size/XS` to `size/L` - only by adding tests that should have been there. 😄 And still the e2e-tests to come....

Reply:
You can move the unit tests to separate PR, that’s a very good idea. Really helpful while reviewing 😅